File: backend/repositories/order_repository.go
package repositories
import (
	"cronos-backend/models"
	"github.com/google/uuid"
	"gorm.io/gorm"
)
type OrderRepository struct {
	DB *gorm.DB
}
func NewOrderRepository(db *gorm.DB) *OrderRepository {
	return &OrderRepository{DB: db}
}
func (r *OrderRepository) Create(order *models.Order) error {
	return r.DB.Create(order).Error
}
func (r *OrderRepository) CreateOrderItem(item *models.OrderItem) error {
	return r.DB.Create(item).Error
}
func (r *OrderRepository) FindByUserID(userID uuid.UUID, page, limit int) ([]models.Order, int64, error) {
	var orders []models.Order
	var total int64
	query := r.DB.Model(&models.Order{}).Where("user_id = ?", userID)
	query.Count(&total)
	offset := (page - 1) * limit
	err := query.Offset(offset).Limit(limit).Order("created_at DESC").
		Preload("OrderItems").Preload("OrderItems.Product").Preload("OrderItems.Product.Images").
		Preload("Payment").Preload("Shipment").
		Find(&orders).Error
	return orders, total, err
}
func (r *OrderRepository) FindAll(page, limit int, status string) ([]models.Order, int64, error) {
	var orders []models.Order
	var total int64
	query := r.DB.Model(&models.Order{})
	if status != "" {
		query = query.Where("status = ?", status)
	}
	query.Count(&total)
	offset := (page - 1) * limit
	err := query.Offset(offset).Limit(limit).Order("created_at DESC").
		Preload("User").Preload("OrderItems").Preload("OrderItems.Product").
		Preload("Payment").Preload("Shipment").
		Find(&orders).Error
	return orders, total, err
}
func (r *OrderRepository) FindByID(id uuid.UUID) (*models.Order, error) {
	var order models.Order
	err := r.DB.Preload("User").Preload("OrderItems").Preload("OrderItems.Product").Preload("OrderItems.Product.Images").
		Preload("Payment").Preload("Payment.MidtransTransaction").
		Preload("Shipment").Preload("Shipment.ShipmentLogs").Preload("Shipment.Courier").
		First(&order, id).Error
	return &order, err
}
func (r *OrderRepository) Update(order *models.Order) error {
	return r.DB.Save(order).Error
}
func (r *OrderRepository) FindBySellerID(sellerID uuid.UUID) ([]models.Order, error) {
	var orders []models.Order
	err := r.DB.Joins("JOIN order_items ON order_items.order_id = orders.id").
		Joins("JOIN products ON products.id = order_items.product_id").
		Where("products.user_id = ?", sellerID).
		Preload("User").Preload("OrderItems").Preload("OrderItems.Product").
		Preload("Payment").Preload("Shipment").
		Group("orders.id").
		Order("orders.created_at DESC").Find(&orders).Error
	return orders, err
}
func (r *OrderRepository) CountByStatus() (map[string]int64, error) {
	type StatusCount struct {
		Status string
		Count  int64
	}
	var results []StatusCount
	err := r.DB.Model(&models.Order{}).Select("status, count(*) as count").Group("status").Find(&results).Error
	if err != nil {
		return nil, err
	}
	counts := make(map[string]int64)
	for _, rc := range results {
		counts[rc.Status] = rc.Count
	}
	return counts, nil
}
func (r *OrderRepository) SumTotalRevenue() (float64, error) {
	var total float64
	err := r.DB.Model(&models.Order{}).Where("status IN ?", []string{"paid", "processing", "shipped", "delivered", "completed"}).
		Select("COALESCE(SUM(total_amount), 0)").Scan(&total).Error
	return total, err
}
func (r *OrderRepository) SumRevenueBySellerID(sellerID uuid.UUID) (float64, error) {
	var total float64
	err := r.DB.Model(&models.OrderItem{}).
		Joins("JOIN products ON products.id = order_items.product_id").
		Joins("JOIN orders ON orders.id = order_items.order_id").
		Where("products.user_id = ? AND orders.status IN ?", sellerID, []string{"paid", "processing", "shipped", "delivered", "completed"}).
		Select("COALESCE(SUM(order_items.subtotal), 0)").Scan(&total).Error
	return total, err
}

File: backend-spring/src/main/java/com/cronos/repository/OrderRepository.java
package com.cronos.repository;

import com.cronos.entity.CronosOrder;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

@Repository
public interface OrderRepository extends JpaRepository<CronosOrder, UUID> {
    Page<CronosOrder> findByUserIdOrderByCreatedAtDesc(UUID userId, Pageable pageable);

    @Query("SELECT o FROM CronosOrder o WHERE (:status IS NULL OR :status = '' OR o.status = :status) ORDER BY o.createdAt DESC")
    Page<CronosOrder> findAllFiltered(@Param("status") String status, Pageable pageable);

    @Query("SELECT DISTINCT o FROM CronosOrder o JOIN OrderItem oi ON oi.orderId = o.id JOIN Product p ON oi.productId = p.id WHERE p.userId = :sellerId ORDER BY o.createdAt DESC")
    List<CronosOrder> findBySellerIdDistinct(@Param("sellerId") UUID sellerId);

    @Query("SELECT o.status, COUNT(o) FROM CronosOrder o GROUP BY o.status")
    List<Object[]> countByStatus();

    @Query("SELECT COALESCE(SUM(o.totalAmount), 0) FROM CronosOrder o WHERE o.status IN :statuses")
    BigDecimal sumTotalRevenue(@Param("statuses") List<String> statuses);

    @Query("SELECT COALESCE(SUM(oi.subtotal), 0) FROM OrderItem oi JOIN Product p ON oi.productId = p.id JOIN CronosOrder o ON oi.orderId = o.id WHERE p.userId = :sellerId AND o.status IN :statuses")
    BigDecimal sumRevenueBySellerIdAndStatuses(@Param("sellerId") UUID sellerId, @Param("statuses") List<String> statuses);
}
